OrthoEvol/Tools/pbs/qsub.py

```diff
@@ -0,0 +1,458 @@
+          '#PBS -l select=1:ncpus=1:mem=6gb'
+        :type chunk_resources:  dict.
+        :return:  A string that will added to the directive list.
+        :rtype:  str.
+        """
+        if chunk_resources is None:
+            chunk_resources = OrderedDict({
+                "select": 1,
+                "ncpus": 1,
+                "mem": "6gb"
+            })
+
+        resource_list = []
+        for k, v in chunk_resources.items():
+            if v is not None:
+                resource_list.append("%s=%s" % (k, v))
+        resource_str = ":".join(resource_list)
+        return resource_str
+
+    def format_template_string(self, code=None, template=None, attributes=None):
+        """
+        Using string Templating, this function takes a dictionary of attributes
+        that are used to format the given code or template file.
+        See https://docs.python.org/3.4/library/string.html#string.Template.
+
+        :param code:  A string of code that is a template.
+        :type code:  str.
+        :param template:  The path to a file that is a template.
+        :type template:  str.
+        :param attributes:  A dictionary of attributes for templating the code/file.
+        :type attributes:  dict.
+        :return:  The template code with the attribute variables filled in.
+        :rtype:  str.
+        """
+
+        if code and attributes is not None:
+            code_template = string.Template(code)
+            code = code_template.substitute(attributes)
+        elif template and attributes is not None:
+            with open(str(template), 'r') as tem:
+                code = tem.read()
+                code_template = string.Template(code)
+                code = code_template.substitute(attributes)
+        else:
+            code = None
+        return code
+
+    def write_template_string(self, code, extension, file=None):
+        """
+        This function writes a code string to a file located in the
+        PBS working directory.  It uses the job_name and the extension to
+        create the file:  <pbs_working_dir>/<job_name>.<extension>
+
+        :param code:  The code that will be written to the file.
+        :type code:   str.
+        :param extension:  The extension of the file.  Usually '.pbs' or '.py'.
+        :type extension:  str.
+        :param file:  A path to the file that will be written.
+        :type file:  str.
+        """
+
+        if file is None:
+            filename = Path(self.pbs_working_dir) / str(self.job_name + extension)
+        else:
+            filename = Path(file)
+        with open(str(filename), 'w') as f:
+            f.write(code)
+
+    def create_header_section(self, file):
+        """
+        Create the header section of the supplied PBS file.
+
+        :param file:  A path to the PBS file.
+        :type file:  str.
+        """
+
+        with open(file, 'a') as f:
+            f.write("# Author:  %s\n" % self.author)
```

Comment:
This could be shortened to something like below...

```python
lines = "# Author:  %s\n# Date Created: %s\n# Project Name: %s\n# Description: %s\n\n"     
with open(file, 'a') as f:
    f.write(lines % (self.author, self.current_date, self.project_name, self.description))
```

Reply:
I like doing it on multiple lines for readability.

Also, I was originally writing it so that you could skip header sections if you wanted to.  Similar to the `create_directives_section` below...

```python
...
if self.author is not None:
    f.write("# Author: %s\n" % self.author)
...
```